File: plugins/extract-go/extract.go
package extractgo

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/shortlink-org/portolan/catalog"
	"github.com/shortlink-org/portolan/plugin"
)

// extract reads a service and answers with one catalog fragment.
//
// A fragment, not a catalog: it carries one context and one service, names
// peers it does not own, and is merged with everything else before anything
// validates it. That is the whole reason the estate does not need a file
// somebody has to own.
func extract(in plugin.Input, opts Options) (plugin.Response, error) {
	b := &plugin.Builder{}
	root := in.Root

	if opts.Context == "" {
		opts.Context = filepath.Base(root)
	}
	if opts.Service == "" {
		opts.Service = filepath.Base(root)
	}

	svcID := serviceID(opts.Context, opts.Service)
	layout := discoverLayout(root)
	if layout.index != nil {
		for _, diagnostic := range layout.index.Diagnostics {
			b.Warn(svcID, diagnostic)
		}
	}
	if opts.Scope != "" {
		layout = layout.scoped(opts.Scope)
	}

	readme := readFile(filepath.Join(root, "README.md"))
	service := catalog.Service{
		ID:       svcID,
		Slug:     opts.Service,
		Name:     firstNonEmpty(opts.ServiceName, markdownTitle(readme), title(opts.Service)),
		Repo:     firstNonEmpty(opts.Repo, modulePath(root)),
		Path:     filepath.ToSlash(root),
		Readme:   readme,
		Provides: []catalog.RpcService{},
		Consumes: []catalog.RpcCall{},
	}

	// Operations live in the application layer and belong to the aggregate the
	// use case sits under, so they are gathered first and handed out below.
	// What exposes each one is read from the transport layer beside it, and the
	// endpoints that came back are where the flows start.
	exposures, endpoints := extractTransport(root, layout, b)
	operations := extractOperations(root, layout, exposures, b)

	for _, aggregateName := range sortedKeys(layout.domains) {
		aggregate, ok := extractAggregate(root, aggregateName, layout.domains[aggregateName], layout, svcID, b)
		if !ok {
			continue
		}
		aggregate.Operations = operations[aggregateName]
		if aggregate.Operations == nil {
			aggregate.Operations = []catalog.Operation{}
		}

		service.Aggregates = append(service.Aggregates, aggregate)
	}

	if len(service.Aggregates) == 0 {
		service.Aggregates = []catalog.Aggregate{}
	}

	for aggregate, ops := range operations {
		if !hasAggregate(service.Aggregates, aggregateID(svcID, slug(aggregate))) {
			b.Warn(svcID, "application slice "+aggregate+" has "+plural(len(ops))+" but there is no matching aggregate domain package")
		}
	}

	// Flows are read last, because a flow names the events the aggregates
	// declare and an event nothing reaches is worth reporting.
	flows, calls := extractFlows(root, flowOptions{
		context:   opts.Context,
		svcID:     svcID,
		service:   opts.Service,
		store:     opts.Store,
		peers:     opts.Peers,
		externals: opts.Externals,
		events:    opts.Events,
	}, layout, endpoints, eventIDs(service.Aggregates), b)
	// Registered execution roots do not require a DDD layout or a deployable
	// scope. Keep the richer domain flow when it already covers a handler.
	covered := map[string]bool{}
	for _, endpoint := range endpoints {
		covered[at(endpoint.source, endpoint.line)] = true
	}
	serviceFlows, serviceCalls := extractServiceFlows(root, opts, b, covered, layout)
	flows = append(flows, serviceFlows...)
	calls = mergeRPCCalls(calls, serviceCalls)
	if len(service.Aggregates) == 0 && len(flows) == 0 && opts.Scope == "" {
		b.Warn(svcID, "no aggregate domain packages or supported execution flows found")
	}
	// What the service calls is read off its flows: the generated client in
	// the tree names the rpc, and a step on it is the call being made.
	service.Consumes = calls

	fragment := catalog.Catalog{
		Contexts: []catalog.BoundedContext{{
			ID:             opts.Context,
			Slug:           opts.Context,
			Name:           firstNonEmpty(opts.ContextName, title(opts.Context)),
			Summary:        opts.ContextSummary,
			Classification: catalog.Classification(opts.Classification),
			Services:       []catalog.Service{service},
		}},
		Defs:  map[string]catalog.TypeDef{},
		Flows: flows,
		Adrs:  []catalog.Adr{},
	}
	spellFromRepository(in, &fragment)

	encoded, err := json.MarshalIndent(fragment, "", "  ")
	if err != nil {
		return plugin.Response{}, err
	}

	b.File(firstNonEmpty(opts.Out, "domain.json"), string(encoded)+"\n")

	return b.Response(), nil
}

// spellFromRepository respells every place in the fragment from the service's
// repository. The reader spells a place as it opened the file, from the
// workspace with the root in front, and matches places on that spelling while
// it reads - a handler a flow already covers, a step's call-site evidence - so
// the respelling happens once, on the finished fragment. In a monorepo the
// workspace is the repository and nothing changes; a fetched copy loses the
// directory it sits in. Function keys (entrypoint, continuesAt, reaches) are
// names, not places, and stay.
func spellFromRepository(in plugin.Input, fragment *catalog.Catalog) {
	for c := range fragment.Contexts {
		services := fragment.Contexts[c].Services
		for s := range services {
			services[s].Path = in.RepositoryPath(services[s].Path)
			for i := range services[s].Consumes {
				call := &services[s].Consumes[i]
				call.Source = in.RepositorySource(call.Source)
				call.Evidence = spelledEvidence(in, call.Evidence)
			}
			for a := range services[s].Aggregates {
				aggregate := &services[s].Aggregates[a]
				for o := range aggregate.Operations {
					aggregate.Operations[o].Source = in.RepositorySource(aggregate.Operations[o].Source)
				}
				for e := range aggregate.Events {
					for v := range aggregate.Events[e].Versions {
						version := &aggregate.Events[e].Versions[v]
						version.Source = in.RepositorySource(version.Source)
					}
				}
				if aggregate.Lifecycle != nil {
					for t := range aggregate.Lifecycle.Transitions {
						transition := &aggregate.Lifecycle.Transitions[t]
						transition.Source = in.RepositorySource(transition.Source)
					}
				}
			}
		}
	}
	for f := range fragment.Flows {
		fragment.Flows[f].Source = in.RepositorySource(fragment.Flows[f].Source)
		spellNodes(in, fragment.Flows[f].Steps)
	}
}

func spellNodes(in plugin.Input, nodes catalog.FlowNodes) {
	for _, node := range nodes {
		switch node := node.(type) {
		case *catalog.Step:
			node.Line = in.RepositorySource(node.Line)
			node.Evidence = spelledEvidence(in, node.Evidence)
			if node.HTTP != nil {
				response := *node.HTTP
				response.Source = in.RepositorySource(response.Source)
				node.HTTP = &response
			}
		case *catalog.Alt:
			for branch := range node.Branches {
				spellNodes(in, node.Branches[branch].Steps)
			}
		case *catalog.Parallel:
			for _, branch := range node.Branches {
				spellNodes(in, branch)
			}
		case *catalog.Loop:
			spellNodes(in, node.Steps)
		}
	}
}

// spelledEvidence is a respelled copy: a list of evidence can be shared by
// the steps of one draft, and each place is respelled once.
func spelledEvidence(in plugin.Input, evidence []catalog.RelationEvidence) []catalog.RelationEvidence {
	if evidence == nil {
		return nil
	}
	out := make([]catalog.RelationEvidence, len(evidence))
	for i, item := range evidence {
		item.Source = in.RepositorySource(item.Source)
		// The enclosing function is named by its package directory as the
		// reader opened it, `<dir>:<Recv>.<Method>`: a place too.
		if item.Rule == "source-function" {
			if dir, name, ok := strings.Cut(item.Symbol, ":"); ok {
				if dir = in.RepositoryPath(dir); dir == "" {
					item.Symbol = name
				} else {
					item.Symbol = dir + ":" + name
				}
			}
		}
		out[i] = item
	}
	return out
}

// eventIDs is every fact the service publishes, which is what a flow reader is
// held to: an event no flow reaches is an event whose publisher this could not
// follow.
func eventIDs(aggregates []catalog.Aggregate) []string {
	var out []string
	for i := range aggregates {
		for j := range aggregates[i].Events {
			out = append(out, aggregates[i].Events[j].ID)
		}
	}

	return out
}

func hasAggregate(aggregates []catalog.Aggregate, id string) bool {
	for i := range aggregates {
		if aggregates[i].ID == id {
			return true
		}
	}

	return false
}

func readFile(path string) string {
	contents, err := os.ReadFile(path)
	if err != nil {
		return ""
	}

	return strings.TrimSpace(string(contents))
}

// modulePath is the module line of go.mod, which is the closest thing the
// source has to a repository name.
func modulePath(root string) string {
	for _, line := range strings.Split(readFile(filepath.Join(root, "go.mod")), "\n") {
		if rest, ok := strings.CutPrefix(strings.TrimSpace(line), "module "); ok {
			return strings.TrimSpace(rest)
		}
	}

	return ""
}

// markdownTitle is the first `# ` heading of a readme, which is what the
// service is actually called.
func markdownTitle(md string) string {
	for _, line := range strings.Split(md, "\n") {
		if heading, ok := strings.CutPrefix(strings.TrimSpace(line), "# "); ok {
			return strings.TrimSpace(heading)
		}
	}

	return ""
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return value
		}
	}

	return ""
}

func plural(n int) string {
	if n == 1 {
		return "1 use case"
	}

	return strconv.Itoa(n) + " use cases"
}

File: plugins/extract-java/src/org/portolan/extract/Extract.java
package org.portolan.extract;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * One Java service in, one fragment out.
 *
 * A fragment, not a catalog: it carries one context and one service, names
 * peers it does not own, and is merged with everything else before anything
 * validates it.
 */
final class Extract {

    private static final Pattern SCM = Pattern.compile("(?s)<scm>.*?<url>\\s*([^<\\s]+)\\s*</url>");

    private Extract() {}

    static void run(Protocol.Input input, Protocol.Options opts, Protocol.Builder b, Path cwd) throws IOException {
        Path root = cwd.resolve(input.root()).normalize();
        java.util.function.Function<Path, String> rel = path -> cwd.relativize(path).toString().replace('\\', '/');

        String context = opts.context.isEmpty() ? root.getFileName().toString() : opts.context;
        String service = opts.service.isEmpty() ? root.getFileName().toString() : opts.service;
        String svcId = Ids.serviceId(context, service);

        Path source = root.resolve(opts.source.isEmpty() ? "src/main/java" : opts.source).normalize();
        Source.Project project = Source.parse(source, rel);
        for (String[] problem : project.broken) {
            b.warn(problem[0], problem[1]);
        }
        if (project.units.isEmpty()) {
            b.warn(svcId, "no Java source under " + rel.apply(source) + ": there is nothing here to read");
        }

        List<Domain.Aggregate> aggregates = Domain.read(project, svcId, root, b);
        Map<String, Events.Found> events = new LinkedHashMap<>();
        for (Domain.Aggregate aggregate : aggregates) {
            events.putAll(Events.read(aggregate, service, b));
        }

        List<Operations.UseCase> useCases = Operations.read(project, b);
        java.util.Set<String> ports = new java.util.LinkedHashSet<>();
        for (Domain.Aggregate aggregate : aggregates) {
            for (com.sun.source.tree.ClassTree port : aggregate.ports) {
                ports.add(port.getSimpleName().toString());
            }
        }
        List<Clients.Client> clients = Clients.read(project, ports, rel, b);
        List<Transport.Endpoint> endpoints = Transport.read(project, rel, b);

        Flows.Options flowOptions = new Flows.Options();
        flowOptions.context = context;
        flowOptions.svcId = svcId;
        flowOptions.service = service;
        flowOptions.store = opts.store;
        flowOptions.peers = opts.peers;
        flowOptions.events = opts.events;
        Flows reader = new Flows(flowOptions, aggregates, useCases, clients, events, b);

        List<Object> flows = new ArrayList<>();
        Map<String, List<String>> exposed = new LinkedHashMap<>();
        for (Transport.Endpoint endpoint : endpoints) {
            flows.add(reader.endpointFlow(endpoint));
            for (String useCase : endpoint.useCases) {
                exposed.computeIfAbsent(useCase, key -> new ArrayList<>()).add(endpoint.id);
            }
        }
        for (Object[] policy : Transport.policies(project)) {
            Map<String, Object> flow = reader.policyFlow((Source.Unit) policy[0], (com.sun.source.tree.ClassTree) policy[1], (com.sun.source.tree.MethodTree) policy[2]);
            if (flow != null) {
                flows.add(flow);
            }
        }

        for (Domain.Aggregate aggregate : aggregates) {
            List<Object> operations = new ArrayList<>();
            for (Operations.UseCase useCase : useCases) {
                if (!belongs(useCase, aggregate)) {
                    continue;
                }
                List<String> by = exposed.get(useCase.id);
                if (by != null) {
                    by.sort(String::compareTo);
                }
                operations.add(Catalog.operation(useCase.id, useCase.kind, useCase.doc, by));
            }
            aggregate.object.put("operations", operations);

            Map<String, Object> lifecycle = Lifecycle.read(aggregate, events, b);
            if (lifecycle != null) {
                aggregate.object.put("lifecycle", lifecycle);
            }

            for (Object event : (List<?>) aggregate.object.get("events")) {
                String id = String.valueOf(((Map<?, ?>) event).get("id"));
                if (!reader.referenced.contains(id)) {
                    b.warn(id, "no flow reaches this event: nothing this extractor could follow publishes it");
                }
            }
        }

        String readme = Files.isRegularFile(root.resolve("README.md")) ? Files.readString(root.resolve("README.md")).strip() : "";

        Map<String, Object> serviceObject = Catalog.map(
                "id", svcId,
                "slug", service,
                "name", firstNonEmpty(opts.serviceName, readmeTitle(readme), Ids.title(service)),
                "repo", opts.repo.isEmpty() ? repo(root) : opts.repo,
                "path", rel.apply(root),
                "readme", readme,
                "provides", new ArrayList<>(),
                "consumes", reader.consumes(),
                "aggregates", aggregates.stream().map(a -> a.object).toList());

        Map<String, Object> contextObject = Catalog.map(
                "id", context,
                "slug", context,
                "name", opts.contextName.isEmpty() ? Ids.title(context) : opts.contextName,
                "summary", opts.contextSummary);
        if (!opts.classification.isEmpty()) {
            contextObject.put("classification", opts.classification);
        }
        contextObject.put("services", List.of(serviceObject));

        Map<String, Object> fragment = Catalog.map(
                "generatedAt", input.generatedAt(),
                "commit", input.commit(),
                "contexts", List.of(contextObject),
                "defs", Catalog.map(),
                "flows", flows,
                "adrs", new ArrayList<>());

        b.file(opts.out.isEmpty() ? "domain.json" : opts.out, Json.write(fragment) + "\n");
    }

    /** A use case belongs to the aggregate its package names, or to the only one there is. */
    private static boolean belongs(Operations.UseCase useCase, Domain.Aggregate aggregate) {
        if (useCase.aggregate.isEmpty()) {
            return false;
        }
        return useCase.aggregate.equals(aggregate.directory);
    }

    private static String readmeTitle(String markdown) {
        for (String line : markdown.split("\n")) {
            if (line.strip().startsWith("# ")) {
                return line.strip().substring(2).strip();
            }
        }
        return "";
    }

    /** The repository url from pom.xml's scm block, spelled the way go.mod spells a module. */
    private static String repo(Path root) throws IOException {
        Path pom = root.resolve("pom.xml");
        if (!Files.isRegularFile(pom)) {
            return "";
        }
        Matcher matcher = SCM.matcher(Files.readString(pom));
        if (!matcher.find()) {
            return "";
        }
        return matcher.group(1)
                .replaceFirst("^git\\+", "")
                .replaceFirst("^(https?|ssh)://", "")
                .replaceFirst("^git@", "")
                .replaceFirst("\\.git$", "")
                .replaceFirst("/$", "");
    }

    private static String firstNonEmpty(String... values) {
        for (String value : values) {
            if (value != null && !value.isEmpty()) {
                return value;
            }
        }
        return "";
    }
}
